packages/editor/src/optimizer/worker.ts

```diff
@@ -289,13 +340,158 @@ const resample = (variation: string): ResampleResult => {
   penroseState = penroseResample({
     ...penroseState!,
     variation,
+    constraintSets: cloneDeep(unpinnedContraintSets),
   });
-  const stepSequenceId = makeNewStepSequence(null, variation);
+  const stepSequenceId = makeNewStepSequence(null, variation, [
+    ...penroseState.varyingValues,
+  ]);
   startOptimize(stepSequenceId);
 
   return taggedOk(MessageTags.Resample, stepSequenceId);
 };
 
+const interact = (data: InteractionRequestData): InteractionResult => {
+  interacting = true;
+
+  const parentSeqInfo = historyInfo.get(data.parentHistoryLoc.sequenceId);
+  if (
+    parentSeqInfo === undefined ||
+    parentSeqInfo.layoutStats.length === 0 ||
+    parentSeqInfo.layoutStats.at(-1)!.cumulativeFrames <=
+      data.parentHistoryLoc.frame
+  ) {
+    return taggedErr(MessageTags.Interaction, {
+      tag: "InvalidHistoryLocError",
+      historyLoc: data.parentHistoryLoc,
+      historyInfo,
+    });
+  }
+
+  const parentValues = historyValues.get(data.parentHistoryLoc.sequenceId)![
+    data.parentHistoryLoc.frame
+  ];
+
+  let currentValues;
+  let stepSeqId: StepSequenceID;
+  if (parentSeqInfo.child === null) {
+    // this is a new interaction
+    const newInitValues = [...parentValues];
+    stepSeqId = makeNewStepSequence(
+      data.parentHistoryLoc,
+      parentSeqInfo.variation,
+      newInitValues,
+    );
+
+    currentValues = historyValues.get(stepSeqId)!.at(-1)!;
+  } else {
+    // we're continuing an interaction
+    stepSeqId = parentSeqInfo.child;
+    currentValues = [...historyValues.get(stepSeqId)!.at(-1)!];
+    historyValues.get(stepSeqId)!.push(currentValues);
```

Comment:
Right now whatever the last set of varying values are in the list is used as the list for the penrose state, so its modified. But this is incredibly silly to worry about from outside `startOptimize`; I'll just change it to be duplicated there.

Reply:
Edit: not quite. These are actually the values that are dragged; if we didn't duplicate them, we'd be erasing the last step. But again, in retrospect, I think the dragging functions should mutate this, they should just return a copy. I'll change that too.

Reply:
> Edit: not quite. These are actually the values that are dragged; if we didn't duplicate them, we'd be erasing the last step. But again, in retrospect, I think the dragging functions should mutate this, they should just return a copy. I'll change that too.

that's my guess. did you mean the dragging functions should return a copy instead of mutating them?

Reply:
fixed.